python/cudf/cudf/pandas/_wrappers/pandas.py

```diff
@@ -1734,6 +1736,88 @@ def _unpickle_obj(pickled_args):
     return obj
 
 
+# Save the original __init__ methods
+_original_Series_init = cudf.Series.__init__
+_original_DataFrame_init = cudf.DataFrame.__init__
+_original_Index_init = cudf.Index.__init__
+
+
+def wrap_init(original_init):
+    @functools.wraps(original_init)
+    def wrapped_init(self, data=None, *args, **kwargs):
+        if is_proxy_object(data):
+            data = data.as_gpu_object()
+            if (
+                isinstance(data, type(self))
+                and len(args) == 0
+                and len(kwargs) == 0
+            ):
+                # This short-circuits the constructor to avoid
+                # unnecessary work when the data is already a
+                # proxy object of the same type.
+                # It is a common case in `cuml` and `xgboost`.
+                # For perf impact see:
+                # https://github.com/rapidsai/cudf/pull/17878/files#r1936469215
+                self.__dict__.update(data.__dict__)
+                return
+        original_init(self, data, *args, **kwargs)
+
+    return wrapped_init
+
+
+@functools.wraps(_original_DataFrame_init)
+def DataFrame_init_(self, data, index=None, columns=None, *args, **kwargs):
+    data_extracted = False
+    if is_proxy_object(data):
+        data = data.as_gpu_object()
+        data_extracted = True
+    elif isinstance(data, list) and len(data) > 0 and is_proxy_object(data[0]):
+        data = [d.as_gpu_object() for d in data]
+    elif (
+        isinstance(data, tuple) and len(data) > 0 and is_proxy_object(data[0])
+    ):
+        data = tuple(d.as_gpu_object() for d in data)
+    elif (
+        isinstance(data, dict)
+        and len(data) > 0
+        and is_proxy_object(next(iter(data.values())))
+    ):
+        data = {k: v.as_gpu_object() for k, v in data.items()}
+
+    if is_proxy_object(index):
+        index = index.as_gpu_object()
+    if is_proxy_object(columns):
+        columns = columns.as_cpu_object()
+    if (
+        (
+            (data is None or (data_extracted and isinstance(data, type(self))))
+            and (index is None)
+            and (columns is None)
+        )
+        and len(args) == 0
+        and len(kwargs) == 0
+    ):
+        self.__dict__.update(data.__dict__)
+        return
+    _original_DataFrame_init(self, data, index, columns, *args, **kwargs)
+
+
+def init_patching():
+    # Replace the __init__ methods with the wrapped versions
+    global \
+        _original_Series_init, \
+        _original_DataFrame_init, \
+        _original_Index_init
+    cudf.Series.__init__ = wrap_init(_original_Series_init)
+    cudf.Index.__init__ = wrap_init(_original_Index_init)
+    cudf.DataFrame.__init__ = DataFrame_init_
+
+
+def initial_setup():
+    init_patching()
```

Comment:
Why split up `initial_setup` and `init_patching`? They seem like they should just be one function.

Reply:
Yes, I was thinking `initial_setup` would serve as a utility for all future initialization-specific stuff. Merged both into one.